Changes in module 'nf-core/seqkit/split2'
--- modules/nf-core/seqkit/split2/main.nf
+++ modules/nf-core/seqkit/split2/main.nf
@@ -2,10 +2,10 @@
     tag "$meta.id"
     label 'process_medium'
 
-    conda "bioconda::seqkit=2.1.0"
+    conda "bioconda::seqkit=2.5.1"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/seqkit:2.1.0--h9ee0642_0' :
-        'biocontainers/seqkit:2.1.0--h9ee0642_0' }"
+        'https://depot.galaxyproject.org/singularity/seqkit%3A2.5.1--h9ee0642_0' :
+        'biocontainers/seqkit==2.5.1--h9ee0642_0' }"
 
     input:
     tuple val(meta), path(reads)

************************************************************
