Changes in module 'nf-core/seqkit/split2'
--- modules/nf-core/seqkit/split2/main.nf
+++ modules/nf-core/seqkit/split2/main.nf
@@ -4,8 +4,8 @@
 
     conda "${moduleDir}/environment.yml"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/seqkit:2.1.0--h9ee0642_0' :
-        'biocontainers/seqkit:2.1.0--h9ee0642_0' }"
+        'https://depot.galaxyproject.org/singularity/seqkit%3A2.5.1--h9ee0642_0' :
+        'biocontainers/seqkit==2.5.1--h9ee0642_0' }"
 
     input:
     tuple val(meta), path(reads)

--- modules/nf-core/seqkit/split2/environment.yml
+++ modules/nf-core/seqkit/split2/environment.yml
@@ -3,4 +3,4 @@
   - bioconda
   - defaults
 dependencies:
-  - bioconda::seqkit=2.1.0
+  - bioconda::seqkit=2.5.1

************************************************************
